docusaurus/docs/create-a-plugin/develop-a-plugin/work-with-data-frames.md

```diff
@@ -0,0 +1,138 @@
+
+If you build a data source plugin, then you'll most likely want to convert a response from an external API to a data frame. Let's look at how to do this.
+
+Let's start with creating a simple data frame that represents a time series. The easiest way to create a data frame is to use the `toDataFrame` function.
+
+```ts
+// Need to be of the same length.
+const timeValues = [1599471973065, 1599471975729];
+const numberValues = [12.3, 28.6];
+
+// Create data frame from values.
+const frame = toDataFrame({
+  name: 'http_requests_total',
+  fields: [
+    { name: 'Time', type: FieldType.time, values: timeValues },
+    { name: 'Value', type: FieldType.number, values: numberValues },
+  ],
+});
+```
+
+:::note
+
+Data frames representing time series contain at least a `time` field and a `number` field. By convention, built-in plugins use `Time` and `Value` as field names for data frames containing time series data.
+
+:::
+
+As you can see from the example, to create data frames like this, your data must already be stored as columnar data. If you already have the records in the form of an array of objects, then you can pass it to `toDataFrame`. In this case, `toDataFrame` tries to guess the schema based on the types and names of the objects in the array. To create complex data frames this way, be sure to verify that you get the schema you expect.
+
+```ts
+const series = [
+  { Time: 1599471973065, Value: 12.3 },
+  { Time: 1599471975729, Value: 28.6 },
+];
+
+const frame = toDataFrame(series);
+frame.name = 'http_requests_total';
+```
+
+## Read values from a data frame
+
+When you're building a panel plugin, the data frames returned by the data source are available from the `data` prop in your panel component.
+
+```ts
+function SimplePanel({ data: Props }) {
+  const frame = data.series[0];
+
+  // ...
+}
+```
+
+Before you start reading the data, think about what data you expect. For example, to visualize a time series you need at least one time field and one number field.
+
+```ts
+const timeField = frame.fields.find((field) => field.type === FieldType.time);
+const valueField = frame.fields.find((field) => field.type === FieldType.number);
+```
+
+Other types of visualizations might need multiple dimensions. For example, a bubble chart that uses three numeric fields: the X-axis, Y-axis, and one for the radius of each bubble. In this case, instead of hard coding the field names, we recommend that you let the user choose the field to use for each dimension.
+
+```ts
+const x = frame.fields.find((field) => field.name === xField);
+const y = frame.fields.find((field) => field.name === yField);
+const size = frame.fields.find((field) => field.name === sizeField);
+
+for (let i = 0; i < frame.length; i++) {
+  const row = [x?.values[i], y?.values[i], size?.values[i]];
+
+  // ...
+}
+```
+
+Alternatively, you can use the `DataFrameView`, which gives you an array of objects that contain a property for each field in the frame.
```

Comment:
follow up - under which circumstances would you use this alternative?

Reply:
tracked in https://github.com/grafana/plugin-tools/issues/435